### PR TITLE
Use duet for asynchrony, in particular for Collector

### PR DESCRIPTION
**Pull request by maffoo**
We are planning to switch to [duet](https://github.com/google/duet) for supporting async code, with one of the reasons being that it is "reentrant" and supports nested invocations of the event loop (`duet.run`). We don't yet have much async code in cirq, but do use it for `Collector` and `PauliSumCollector` and this can lead to problems if running in a context where the asyncio event loop is already running (#4004). The solution proposed there is to patch asyncio to make it reentrant, using the `nest-asyncio` library, but this seems like something we should not rely on. Instead, I propose that we make the switch to duet.

The main question I have is whether we need to continue to support `Collector.collect_async` and `Sampler.*_async` methods being asyncio coroutines, or whether we can just switch them to be duet coroutines. In this PR as it stands I've just switched things, but if we want to support asyncio we would need to name the duet coroutine functions differently and then wrap them into an asyncio-compatible form for the existing methods.

### CODE REVIEW

**Comment by 95-martin-orion** [comment]
Please fix the failing tests, then re-request review.

**Comment by maffoo** [comment]
Hm, we don't have a version of duet for 3.6, and it'd be quite hard to do because of some asnyc/await language features that landed in 3.7. @balopat, what is the plan for dropping python 3.6 support?

**Comment by balopat** [comment]
This is blocked on #3574, which is blocked on google3 having their default python runtime upgraded to 3.7. 
If we want to move forward with duet for the OSS community (I think that this is great!) then we have to think about workarounds until that upgrade is done which might take a year still. A workaround design might mean to offer the current async flow on 3.6 and duet on 3.7 somehow.

**Comment by balopat** [comment]
> This is blocked on #3574, which is blocked on google3 having their default python runtime upgraded to 3.7.
> If we want to move forward with duet for the OSS community (I think that this is great!) then we have to think about workarounds until that upgrade is done which might take a year still. A workaround design might mean to offer the current async flow on 3.6 and duet on 3.7 somehow.

One idea: 

We could add a cirq-duet module that would use duet for the collector. 
People who have python 3.7+ could use that module, the ones using 3.6 can still use the old collector, on which we could add even a deprecation warning...

WDYT, @maffoo?

**Comment by mpharrigan** [comment]
Do we have an update on timeline for Python 3.7+?

**Comment by maffoo** [comment]
Actually, @fedimser is working on adding python 3.6 support to duet (https://github.com/google/duet/pull/23) which will let us move forward with these async changes without waiting for 3.7 support in colab etc.

**Comment by maffoo** [comment]
This is ready for review. @balopat, @mpharrigan

File: cirq-core/cirq/conftest.py
Comment (mpharrigan) [inline cirq-core/cirq/conftest.py:17]: nit: It should

File: cirq-core/cirq/testing/__init__.py
Comment (mpharrigan) [inline cirq-core/cirq/testing/__init__.py:18]: This is technically a breaking change. I suspect no one externally has been using our async testing helper code? but we should make a note of this for the release

File: cirq-core/cirq/work/collector.py
Comment (mpharrigan) [inline cirq-core/cirq/work/collector.py:134]: lovely

File: cirq-core/cirq/work/collector.py
Comment (mpharrigan) [inline cirq-core/cirq/work/collector.py:169]: what is going on here? Is this a form of type annotation I've never seen before or does this actually change the object that has been created?

File: cirq-core/cirq/work/collector.py
Comment (mpharrigan) [inline cirq-core/cirq/work/collector.py:183]: what is try .. else? why not put results.add inside the try?

File: cirq-core/cirq/work/work_pool.py
Comment (mpharrigan) [inline cirq-core/cirq/work/work_pool.py:20]: this seems like something someone may be relying on in cirq user land?

**Review by mpharrigan — COMMENTED** [review]
I think the actual code changes in this PR are very straightforward. I asked some clarifying questions and posed some nits.

File: cirq-core/cirq/work/collector.py
Comment (maffoo) [reply cirq-core/cirq/work/collector.py:184]: The `else` clause executes if the `try` block does not raise any exceptions. I wanted to make clear that the `except` clause is _only_ catching exceptions from the await call. If `results.add` were part of the `try` clause then the except would also handle errors from that; `results.add` shouldn't raise errors, but this is just to be safe.

File: cirq-core/cirq/work/work_pool.py
Comment (maffoo) [reply cirq-core/cirq/work/work_pool.py:20]: It's certainly possible. I can restore this and deprecate it instead.

File: cirq-core/cirq/testing/__init__.py
Comment (maffoo) [reply cirq-core/cirq/testing/__init__.py]: Brought this back since it's needed to test the work pool, which is now deprecated but not removed.

File: cirq-core/cirq/work/collector.py
Comment (mpharrigan) [reply cirq-core/cirq/work/collector.py:170]: ok, glad it's just a type hint. my slight preference would be for the type annotation at the cost of some repetition but I defer to you

File: cirq-core/cirq/work/collector.py
Comment (mpharrigan) [reply cirq-core/cirq/work/collector.py:184]: thanks for explaining

**Review by mpharrigan — APPROVED** [review]

File: cirq-core/cirq/work/collector.py
Comment (maffoo) [reply cirq-core/cirq/work/collector.py]: Switched to type annotation. It still _just_ fits on one line :-)

**Review by dstrain115 — APPROVED** [review]
In general, it might be good to put more line comments in where duet is used, since it will be unfamiliar to the majority of developers, but I don't have any other objections to pulling this in.

**Pull request merged**
